GVFS/GVFS.Common/Maintenance/GitMaintenanceStep.cs

```diff
@@ -0,0 +1,172 @@
+﻿using GVFS.Common.Git;
+using GVFS.Common.Tracing;
+using System;
+using System.IO;
+
+namespace GVFS.Common.Maintenance
+{
+    public abstract class GitMaintenanceStep
+    {
+        public const string ObjectCacheLock = "git-maintenance-step.lock";
+        private readonly object gitProcessLock = new object();
+
+        public GitMaintenanceStep(GVFSContext context, GitObjects gitObjects, bool requireObjectCacheLock)
+        {
+            this.Context = context;
+            this.GitObjects = gitObjects;
+            this.RequireObjectCacheLock = requireObjectCacheLock;
+        }
+
+        public abstract string Area { get; }
+
+        protected GVFSContext Context { get; }
+        protected GitObjects GitObjects { get; }
+        protected GitProcess GitProcess { get; private set; }
+        protected bool Stopping { get; private set; }
+        protected bool RequireObjectCacheLock { get; }
+
+        public void Execute()
+        {
+            try
+            {
+                if (this.RequireObjectCacheLock)
+                {
+                    using (FileBasedLock cacheLock = GVFSPlatform.Instance.CreateFileBasedLock(
+                        this.Context.FileSystem,
+                        this.Context.Tracer,
+                        Path.Combine(this.Context.Enlistment.GitObjectsRoot, ObjectCacheLock)))
+                    {
+                        if (!cacheLock.TryAcquireLock())
+                        {
+                            this.Context.Tracer.RelatedInfo(this.Area + ": Skipping work since another process holds the lock");
+                            return;
+                        }
+
+                        this.CreateProcessAndRun();
+                    }
+                }
+                else
+                {
+                    this.CreateProcessAndRun();
+                }
+            }
+            catch (IOException e)
+            {
+                this.Context.Tracer.RelatedWarning(
+                    metadata: this.CreateEventMetadata(e),
+                    message: this.Area + ": IOException while running action: " + e.Message,

```

Comment:
Seems inconsistent how `this.Area` is being used.  In some places below it is only in the message and others it is in the metadata field `"Area"`.  Here it is in both. I think we would just need it in the metadata field and leave it out of the message.

Reply:
I audited instances of `CreateEventMetadata()` and removed the duplicate areas.